tmc-plugin/src/fi/helsinki/cs/tmc/spyware/eventsources/OutputActionCaptor.java

```diff
@@ -0,0 +1,267 @@
+        for (IOProvider provider : Lookup.getDefault().lookupAll(IOProvider.class)) {
+            if (provider.getName().equals(NAME)) {
+                continue;
+            }
+
+            InputOutput a = provider.getIO(string, bln);
+//            a.setFocusTaken(true);
+            return new TmcInputOutputProxy(a);
+        }
+        throw new RuntimeException("Failed to get IO. Please contact TestMyCode authors/teacher");
+
+    }
+
+    /**
+     * Proxies access to InputOutput from the default provider. Used as one
+     * event source for spyware module.
+     */
+    private static final class TmcInputOutputProxy implements InputOutput {
+
+        private final InputOutput original;
+        private final Exercise exercise;
+
+        public TmcInputOutputProxy(InputOutput original) {
+            this.original = original;
+            this.exercise = getExercise(getChangedFile());
+        }
+
+        private Exercise getExercise(@NullAllowed FileObject obj) {
+            if (obj == null) {
+                return null;
+            }
+            ProjectMediator pm = ProjectMediator.getInstance();
+            TmcProjectInfo project = pm.tryGetProjectOwningFile(obj);
+            return pm.tryGetExerciseForProject(project, courseDb);
+        }
+
+        /**
+         * Returns {@link FileObject} representing the last active file for each
+         * event.
+         */
+        private FileObject getChangedFile() {
+            JTextComponent jtc = EditorRegistry.lastFocusedComponent();
+            if (jtc != null) {
+                Document d = jtc.getDocument();
+                Object fileObj = d.getProperty(NbEditorDocument.StreamDescriptionProperty);
+                if (fileObj instanceof DataObject) {
+                    DataObject changedDataObject = (DataObject) fileObj;
+                    return changedDataObject.getPrimaryFile();
+                }
+            }
+            return null;
+        }
+
+        @Override
+        public OutputWriter getOut() {
+            return original.getOut();
+        }
+
+        @Override
+        public Reader getIn() {
+            final Reader reader = original.getIn();
+            return new TmcReader(reader);
+        }
+
+        /**
+         * Proxies access to Reader, allowing us to sniff the interactions.
+         */
+        private final class TmcReader extends Reader {
+
+            private final Reader original;
+
+            public TmcReader(Reader original) {
+                this.original = original;
+            }
+
+            @Override
+            public int read(char[] cbuf, int off, int len) throws IOException {
+                return original.read(cbuf, off, len);
+            }
+
+            private void emit(char c) {
+                for (Listener listener : listeners) {
+                    listener.input(exercise, c);
+                }
+            }
+
+            @Override
+            public int read() throws IOException {
```

Comment:
Good point, fixd.Thou it seems that pastes are considered to be individual characters typed in really fast pace.